### PR TITLE
Add property tests for multi_index

### PR DESCRIPTION
**Pull request by ihnorton**
Adds two property-based tests for `multi_index` using [Hypothesis](https://hypothesis.readthedocs.io/en/latest/)
- test `Array.multi_index` against direct query on arbitrary generated ranges
- test `Array.multi_index` on indexes generated from acceptable input spec

### CODE REVIEW

File: tiledb/tests/test_multi_index-hp.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    return res
+
+
+# Compound strategies to build valid inputs for multi_index
+tuple_indexer = st.one_of(
+    st.none(),
+    st.integers(),
+)
+subindex_obj = st.one_of(
+    st.none(),
```
Comment (gsakkis) [inline tiledb/tests/test_multi_index-hp.py:45]: Why is this needed? Afaik we don't support indexing by None

File: requirements_dev.txt [hunk 2]
```diff
@@ -9,4 +9,6 @@ pandas ; python_version > '3.5'
 psutil
 pyarrow
 pytest
+hypothesis
+hypothesis[numpy]
```
[the remark this answers is not in the document]
Comment (ihnorton) [reply requirements_dev.txt:13]: Yes

File: tiledb/tests/common.py [hunk 3]
```diff
@@ -14,6 +14,8 @@
 import numpy as np
 from numpy.testing import assert_equal, assert_array_equal
 
+from .hpcommon import *
```
[the remark this answers is not in the document]
Comment (ihnorton) [reply tiledb/tests/common.py]: Updated: renamed to `tiledb.tests.strategies` and fixed imports

File: tiledb/tests/test_multi_index-hp.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+from numpy.testing import assert_array_equal
+
+from pathlib import Path
+import warnings
+
+import pytest
+from tiledb.tests.common import checked_path
+from tiledb.tests.common import bounded_ntuple, ranged_slices
+
+from hypothesis import given, assume, settings
```
[the remark this answers is not in the document]
Comment (ihnorton) [reply tiledb/tests/test_multi_index-hp.py:16]: Yes, I use it sometimes to run more exhaustive than the default 100

File: tiledb/tests/test_multi_index-hp.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+    res = {k: v[0].view(array.attr(0).dtype) for k, v in q.results().items()}
+    return res
+
+
+# Compound strategies to build valid inputs for multi_index
+tuple_indexer = st.one_of(
+    st.none(),
+    st.integers(),
+)
```
[the remark this answers is not in the document]
Comment (ihnorton) [reply tiledb/tests/test_multi_index-hp.py]: removed

File: tiledb/tests/test_multi_index-hp.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+subindex_obj = st.one_of(
+    st.none(),
+    st.integers(),
+    ranged_slices(),
+    tuple_indexer,
+)
+
+index_obj = st.one_of(st.tuples(subindex_obj), st.tuples(st.lists(subindex_obj)))
```
[the remark this answers is not in the document]
Comment (ihnorton) [reply tiledb/tests/test_multi_index-hp.py]: updated

File: tiledb/tests/test_multi_index-hp.py (hunk 1 above)
Comment (ihnorton) [reply tiledb/tests/test_multi_index-hp.py]: removed

**Review by gsakkis — APPROVED** [review]
